### PR TITLE
Return proper error code in precision loss edge case in exchange rates service

### PR DESCRIPTION
**Pull request by gakonst**
Fixes https://github.com/interledger-rs/interledger-rs/issues/499

### CODE REVIEW

File: crates/interledger-service-util/src/exchange_rates_service.rs [hunk 1]
```diff
@@ -146,7 +146,7 @@ where
                     // When it gets cast to a u64, it will end up being 0.
                     if outgoing_amount != 0.0 && outgoing_amount as u64 == 0 {
                         return Box::new(err(RejectBuilder {
-                            code: ErrorCode::F08_AMOUNT_TOO_LARGE,
+                            code: ErrorCode::R01_INSUFFICIENT_SOURCE_AMOUNT,
```
Comment (emschwartz) [inline crates/interledger-service-util/src/exchange_rates_service.rs]: I don't think that's the right error. The problem here is that the amount is bigger than the max u64. You'd definitely want to communicate to the sender that they need to send a smaller amount in order for the packet to go through

Comment (gakonst) [reply crates/interledger-service-util/src/exchange_rates_service.rs]: PR was made based on: https://github.com/interledger-rs/interledger-rs/issues/499#issuecomment-550015494

Comment (emschwartz) [reply crates/interledger-service-util/src/exchange_rates_service.rs]: Ah my mistake, this is when it gets rounded down to 0. Maybe R01 is the right error in this case

Comment (gakonst) [reply crates/interledger-service-util/src/exchange_rates_service.rs]: @emschwartz So the way Rust handles f64 being cast to u64 when it doesn't fit, is that it gets zero'ed out. 

https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=c81a1a7a2cc882ea787fc7e7fe6b01d3

that said, insufficient source amount is definitely not correct, as it's the result of overflow. My follow up commit, made that an `INVALID_AMOUNT` error, which seems like it makes more sense.

Comment (emschwartz) [reply crates/interledger-service-util/src/exchange_rates_service.rs]: Ok if it's the result of an overflow, `F08_AMOUNT_TOO_LARGE` makes more sense to me than `F03_INVALID_AMOUNT`. The problem is not that the sender sent an invalid number, it's that they sent enough to cause us to have an overflow and they should try again with a smaller amount. `F08` seems more helpful

File: crates/interledger-service-util/src/exchange_rates_service.rs [hunk 2]
```diff
@@ -146,7 +146,7 @@ where
                     // When it gets cast to a u64, it will end up being 0.
                     if outgoing_amount != 0.0 && outgoing_amount as u64 == 0 {
```
Comment (emschwartz) [inline crates/interledger-service-util/src/exchange_rates_service.rs]: @gakonst #499 points at the issue when the amount gets rounded down to 0 because it's less than 1. I think there should be two checks here.

```
if outgoing_amount != 0.0 && outgoing_amount as u64 == 0 {
  if outgoing_amount < 1.0 {
    // user wanted to send a positive value but it got rounded down to 0
    return R01
  } else {
    // amount that arrived was too large for us to forward
    return F08
  }
}
```

}

Comment (gakonst) [reply crates/interledger-service-util/src/exchange_rates_service.rs]: OK I agree with this

File: crates/interledger-service-util/src/exchange_rates_service.rs [hunk 3]
```diff
@@ -145,13 +145,29 @@ where
...
+                            // user wanted to send a positive value but it got rounded down to 0
+                            (
+                                ErrorCode::R01_INSUFFICIENT_SOURCE_AMOUNT,
+                                format!(
+                                    "Could not cast to f64, amount too small: {}",
+                                    outgoing_amount
+                                ),
+                            )
+                        } else {
+                            // amount that arrived was too large for us to forward
```
Comment (emschwartz) [inline crates/interledger-service-util/src/exchange_rates_service.rs]:
```suggestion
                            // amount that arrived was too large and the overflowing float was
                            // converted to 0 when it was cast to a u64
```

File: crates/interledger-service-util/src/exchange_rates_service.rs [hunk 4]
```diff
@@ -145,13 +145,29 @@ where
...
+                            (
+                                ErrorCode::F08_AMOUNT_TOO_LARGE,
```
Comment (kincaidoneil) [inline crates/interledger-service-util/src/exchange_rates_service.rs]: AFAIK, this case would probably occur when a user accidentally fat fingers wayyyyy too much _and_ this connector doesn't have any max packet amount set (which would already be caught earlier in the pipeline), right?

By returning an F08 error without any associated max packet data, their STREAM sender will retry, decreasing the packet amount by a factor of 2. Eventually, it'd try an amount that'd convert below the max u64, but still a massive amount (!), and that poor user might lose money if they have automatic settlement or something.

Alternatively, if we returned an F03, their STREAM sender wouldn't have that behavior, which might provide a better sanity check in this (admittedly low probability) case.

TL;DR maybe this should be a "final final" error instead of a "retryable final" error.

Comment (emschwartz) [reply crates/interledger-service-util/src/exchange_rates_service.rs]: I do not think this is the place to handle such an issue. There should be enough confirmation steps or sanity checks in the sender's client to prevent them sending a ridiculous amount.
